src/scheduler/job_info.c

```diff
@@ -949,15 +837,248 @@ query_jobs(status *policy, int pbs_sd, queue_info *qinfo, resource_resv **pjobs,
+			sprintf(log_buffer, "pbs_selstat failed: %s (%d)", errmsg, pbs_errno);
+			schdlog(PBSEVENT_SCHED, PBS_EVENTCLASS_JOB, LOG_NOTICE, "job_info", log_buffer);
+		}
+		return pjobs;
+	}
+
+	/* count the number of new jobs */
+	cur_job = jobs;
+	while (cur_job != NULL) {
+		num_jobs++;
+		cur_job = cur_job->next;
+	}
+	num_new_jobs = num_jobs;
+
+	/* if there are previous jobs, count those too */
+	num_prev_jobs = count_array((void **) pjobs);
+	num_jobs += num_prev_jobs;
+
+
+	/* allocate enough space for all the jobs and the NULL sentinal */
+	resresv_arr = (resource_resv **) realloc(pjobs, sizeof(resource_resv*) * (num_jobs + 1));
+
+	if (resresv_arr == NULL) {
+		log_err(errno, __func__, MEM_ERR_MSG);
+		pbs_statfree(jobs);
+		return NULL;
 	}
-	resresv_arr[i] = NULL;
+	resresv_arr[num_prev_jobs] = NULL;
+
+	j = 0;
+	chunk_size = num_new_jobs / num_threads;
+	chunk_size = (chunk_size > 1024) ? chunk_size : 1024;
+	chunk_size = (chunk_size < 8192) ? chunk_size : 8192;
+	while (num_new_jobs > 0) {
+		tdata = malloc(sizeof(th_data_query_jinfo));
+		if (tdata == NULL) {
+			log_err(errno, __func__, MEM_ERR_MSG);
+			th_err = 1;
+			break;
+		}
+		tdata->err = 0;
+		tdata->jobs = jobs;
+		tdata->oarr = NULL; /* Will be filled by the thread routine */
+		tdata->sinfo = qinfo->server;
+		tdata->qinfo = qinfo;
+		tdata->pbs_sd = pbs_sd;
+		tdata->policy = policy;
+		tdata->sidx = j;
+		j += chunk_size;
+		tdata->eidx = j - 1;
+		task = malloc(sizeof(th_task_info));
+		if (task == NULL) {
+			free(tdata);
+			log_err(errno, __func__, MEM_ERR_MSG);
+			th_err = 1;
+			break;
+		}
+		task->task_id = num_tasks;
+		task->task_type = TS_QUERY_JOB_INFO;
+		task->thread_data = (void *) tdata;
+
+		pthread_mutex_lock(&work_lock);
+		ds_enqueue(work_queue, (void *) task);
+		pthread_cond_signal(&work_cond);
+		pthread_mutex_unlock(&work_lock);
+
+		num_tasks++;
+		num_new_jobs -= chunk_size;
+	}
+	jinfo_arrs_tasks = malloc(num_tasks * sizeof(resource_resv **));
+	if (jinfo_arrs_tasks == NULL) {
+		log_err(errno, __func__, MEM_ERR_MSG);
+		th_err = 1;
+	}
+	/* Get results from worker threads */
+	for (i = 0; i < num_tasks;) {
+		pthread_mutex_lock(&result_lock);
+		while (ds_queue_is_empty(result_queue))
+			pthread_cond_wait(&result_cond, &result_lock);
+		pthread_mutex_unlock(&result_lock);
+		while (!ds_queue_is_empty(result_queue)) {
+			task = (th_task_info *) ds_dequeue(result_queue);
+			tdata = (th_data_query_jinfo *) task->thread_data;
+			if (tdata->err)
+				th_err = 1;
+			jinfo_arrs_tasks[task->task_id] = tdata->oarr;
+			free(tdata);
+			free(task);
+			i++;
+		}
+	}
+	if (th_err) {
+		pbs_statfree(jobs);
+		free_resource_resv_array(resresv_arr);
+		return NULL;
+	}
+	/* Assemble job info objects from various threads into the resresv_arr */
+	for (i = 0, jidx = num_prev_jobs; i < num_tasks; i++) {
+		if (jinfo_arrs_tasks[i] != NULL) {
+			for (j = 0; (resresv = jinfo_arrs_tasks[i][j]) != NULL; j++) {
+				resresv_arr[jidx++] = resresv;
```

Comment:
Why bother with the resresv variable?  You only use it once and setting it in the while portion of the for loop is strange

Reply:
I use it twice right? once to check that resresv != NULL, and once to add it to the resresv_arr.

Reply:
You don't really use it twice.  When you check resresv != NULL, you're also setting resresv.  You are using job_arr_tasks[i][j] there as well.  I'd just use job_arrs_tasks[i][j] twice.